
We apply Prokounter to a few datasets. 
```{r}
require(metagenomeSeq)
require(matrixStats)
require(metagenomeSeq)
require(portra50)
require(matrixStats)
require(MASS)
require(prokounter)
require(breakaway)

source('~/Documents/research/StatisticalAnalysisProjects/NSPSU/utils_prokounter.R')
source('~/Documents/research/intensityBias/src/otherDivsCalls.R')

pred.plot <- function(df, g2ft, fit, plt.knots=FALSE, alpha=.1, methd="ss",... ){
  library(scales)
  
  if(methd=="ss"){
    lrvec <- seq(min(df$lr), max(df$lr), .1)
    newd = data.frame( "lr"= rep(lrvec,
                                 each=length(g2ft) ),
                       "g"= rep( g2ft,
                                 each=length( lrvec)
                       )
    )
    ng <- length( df$g )
    cc <- getColsForFactors( factor(df$g), ... )
    trans.cols <- scales:::alpha( cc$cols, alpha=alpha )
    ltv <- predict( fit, newdata=newd[,"lr",drop=FALSE], include=c( "1", "lr"), se.fit=TRUE);
    plot( df$lm ~ df$lr, col = trans.cols, xlab = "Log2 Recovered Abundance", ylab = "Log2 Num. Taxa", pch=19 )
    #plot( df$lm ~ df$lr, col = cc$cols, xlab = "Log2 Recovered Abundance", ylab = "Log2 Num. Taxa" )
    #smoothScatter( df$lm ~ df$lr, col = cc$cols, xlab = "Log2 Recovered Abundance", ylab = "Log2 Num. Taxa" )
    abline( h=0 )
    o <- order( newd$lr)
    lines( ltv$fit[o] ~ newd$lr[o], col = 'red', cex=.5 )
    lines( (ltv$fit[o]+1.96*ltv$se[o]) ~ newd$lr[o], col = 'red', cex=.5, lty=2 )
    lines( (ltv$fit[o]-1.96*ltv$se[o]) ~ newd$lr[o], col = 'red', cex=.5, lty=2 )
    if(plt.knots)
      points( df$lm[fit$id.basis] ~ df$lr[ fit$id.basis ], pch=23, col = 'blue', bg='blue' )
  }
  
  else {
    lrvec <- seq(min(df$lr), max(df$lr), .1)
    newd = data.frame( "lr"= lrvec )
    ltv <- predict( fit, newdata=newd,se=TRUE ); 
    cc <- getColsForFactors( factor(df$g), ... )
    trans.cols <- scales:::alpha( cc$cols, alpha=alpha )
    plot( df$lm ~ df$lr, col = trans.cols, xlab = "Log2 Recovered Abunadnce", ylab = "Log2 Num. Taxa", pch=19 )
    abline( h=0 )
    o <- order( newd$lr )
    lines( ltv$fit[o] ~ newd$lr[o], col = 'red', cex=.5 )
    lines( (ltv$fit[o]+1.96*ltv$se[o]) ~ newd$lr[o], col = 'red', cex=.5, lty=2 )
    lines( (ltv$fit[o]-1.96*ltv$se[o]) ~ newd$lr[o], col = 'red', cex=.5, lty=2 )
  }
}
```



# TREND MODELS
```{r}
dfs <-list()
genus.var <- list()

dfs$psu.fst99 <- readRDS(file="~/Documents/research/intensityBias/data/PIH_CSF_100_Analysis_Jan2021/PIH_CSF_First100_data2_Qiime1_Jan2021/PIH_CSF_First100_99_SILVA132.RDS")
dfs$psu.fst97 <- readRDS(file="~/Documents/research/intensityBias/data/PIH_CSF_100_Analysis_Jan2021/PIH_CSF_First100_data2_Qiime1_Jan2021/PIH_CSF_First100_97_SILVA132.RDS")
dfs$dada2 <- readRDS(file="~/Documents/research/intensityBias/data/PIH_CSF_100_Analysis_Jan2021/PIH_CSF_First100_data2_Qiime2_Jan2021/PIH_CSF_First100_Dada2_SILVA132.RDS")
genus.var$psu.fst99 <- "Genus"
genus.var$psu.fst97 <- "Genus"
genus.var$dada2 <- "Genus"

load( file="./data/wastewater/wastewater.rdata" ) #loads wastewaterMRobj
dfs$ww <- convertMRtoES(wastewaterMRobj)
genus.var$ww <- "genus"

dfs$alm <- readRDS( file="./data/alm.AF.RDS" )
genus.var$alm <- "taxonomy6"


dfs2 <- readRDS( file="~/Documents/research/intensityBias/data/cycvardat.RDS" )
dfs$pseudo99 <- dfs2$FST.99
genus.var$pseudo99 <- "Genus"
dfs$pseudo97 <- dfs2$FST.97
genus.var$pseudo97 <- "Genus"
dfs$pseudodada <- dfs2$Dada2
genus.var$pseudodada <- "Genus"



nx <- names(dfs); names(nx) <- nx
pk.ss <- lapply( nx, function(y){
  obj <- dfs[[y]]
  gvar <- genus.var[[y]]
  gs <- fData(obj)[[gvar]]
  pk <- prokounter::getProkounterTrends( exprs(obj), genus=as.character(gs), fit.proc = "ss", plt=FALSE )
  pk
}  )
pk.lo <- lapply( nx, function(y){
  obj <- dfs[[y]]
  gvar <- genus.var[[y]]
  gs <- fData(obj)[[gvar]]
  pk <- prokounter::getProkounterTrends( exprs(obj), genus=as.character(gs), fit.proc = "lo", plt=FALSE )
  pk
}  ) 

LEGEND.NAMES <- c( "PIH100 FST.99", "PIH100 FST.97", "PIH100 Dada2", 
                   "Wastewater FST.97", "TS FST.97", "PDil FST.99", "PDil FST.97", "PDil Dada2"  )
pdf( file="./results/images/lotrends.pdf", useDingbats = FALSE )
op <-par(mfrow=c(2,3), pty='s')
cntr <- 1
for( x in names(dfs) ){
  obj <- dfs[[x]]
  gvar <- genus.var[[x]]
  gs <- fData(obj)[[gvar]]
  pk <- pk.lo[[x]]
  pred.plot( df = pk$df, fit=pk$fit, g2ft = pk$ug, methd = "lo", alpha = .25 )
  legend( "topleft", LEGEND.NAMES[cntr],  bty="n" )
  cntr <- cntr + 1
}
dev.off()
par(op)

pdf( file="./results/images/sstrends.pdf", useDingbats = FALSE )
op <-par(mfrow=c(2,3), pty='s')
cntr <- 1
for( x in names(dfs) ){
  obj <- dfs[[x]]
  gvar <- genus.var[[x]]
  gs <- fData(obj)[[gvar]]
  pk <- pk.ss[[x]]
  pred.plot( df = pk$df, fit=pk$fit, g2ft = pk$ug, alpha=.25 )
  legend( "topleft", LEGEND.NAMES[cntr],  bty="n" )
  cntr <- cntr + 1
}
dev.off()
par(op)
  

```

Adding MSD as well. 
```{r}
library(msd16s)
msd <- convertMRtoES(msd16s)

pdf( file="./results/images/MSD_numTaxa_RAplot_SS.pdf" )
op <- par(mfrow=c(1,2), pty='s')
pkmsd <- prokounter::getProkounterTrends( mat=exprs(msd), genus=as.character(fData(msd)$genus), plt=TRUE )
dev.off()

# we couldn't manage to fit a loess to the whole data because of "vector memory exhausted".
pdf( file="./results/images/MSD_numTaxa_RAplot_LO.pdf" ) #
op <- par(mfrow=c(1,2), pty='s')
pkmsd <- prokounter::getProkounterTrends( mat=exprs(msd), genus=as.character(fData(msd)$genus), fit.proc="lo", plt=TRUE, span=.15  )
dev.off()
```